simulator/src/main/cook/sim/reporting/groups.clj

```diff
@@ -0,0 +1,83 @@
+(ns cook.sim.reporting.groups
+  (:require [clojure.pprint :as pprint]
+            [datomic.api :as d]
+            [cook.sim.database :as data]
+            [cook.sim.util :as u]))
+
+
+(defn find-groups
+  "Given a collection of job-results, returns only the jobs with groups,
+  in a map keyed by group name."
+  [job-results]
+  (dissoc (group-by :group job-results) nil))
+
+(defn time-of-first-request
+  "Given a collection of job results, return the earliest time that any was requested."
+  [job-results]
+  (->> job-results (map :requested-at) (apply min)))
+
+(defn sorted-start-times
+  "Returns the start times for jobs that HAVE start times, ordered"
+  [job-results]
+  (->> job-results (map :started-at) (filter identity) sort))
+
+(defn wait-for-first
+  "Given a collection of job results, return the time between when the first one
+  was requested and the first time any of them was scheduled."
+  [job-results]
+  (let [time (-> job-results sorted-start-times first)]
+    (when time (- time (time-of-first-request job-results)))))
+
+(defn wait-for-last
+  "Given a collection of job results, return the time between when the first one
+  was requested and the latest time any of them was scheduled."
+  [job-results]
+  (let [time (-> job-results sorted-start-times last)]
+    (when time (- time (time-of-first-request job-results)))))
+
+(defn runtime
+  "Given a collection of job results, return the time between when the first moment
+  any of them was started requested and the last time when one finished."
+  [job-results]
+  (let [first-start (-> job-results sorted-start-times first)
+        last-finish (->> job-results (map :finished-at) (filter identity) sort last)]
+    (when (and first-start last-finish)
+      (- last-finish first-start))))
+
+(defn average-by-group
+  "Separates a collection of job-results into groups, performs the given summary
+  function on the jobs in each group, and returns the average of these results
+  across all groups."
+  [summary-fn job-results]
+  (let [groups (find-groups job-results)
+        total-ms (->> groups
+                      vals
+                      (map summary-fn)
+                      (filter identity)
```

Comment:
`(remove nil?)` makes the intent clearer of removing the nil values instead of `(filter identity)`